fml/message_loop_task_queues_unittests.cc

```diff
@@ -88,6 +117,46 @@ TEST(MessageLoopTaskQueue, PreserveTaskOrdering) {
   }
 }
 
+TEST(MessageLoopTaskQueue, RegisterTasksOnMergedQueuesPreserveTaskOrdering) {
+  auto task_queue = fml::MessageLoopTaskQueues::GetInstance();
+  auto platform_queue = task_queue->CreateTaskQueue();
+  auto raster1_queue = task_queue->CreateTaskQueue();
+  auto raster2_queue = task_queue->CreateTaskQueue();
+  int test_val = 0;
+
+  // order 0 in raster1_queue
+  task_queue->RegisterTask(
+      raster1_queue, [&test_val]() { test_val = 0; }, fml::TimePoint::Now());
+
+  // order 1 in platform_queue
+  task_queue->RegisterTask(
+      platform_queue, [&test_val]() { test_val = 1; }, fml::TimePoint::Now());
+
+  // order 2 in raster2_queue
+  task_queue->RegisterTask(
+      raster2_queue, [&test_val]() { test_val = 2; }, fml::TimePoint::Now());
+
+  task_queue->Merge(platform_queue, raster1_queue);
+  ASSERT_TRUE(task_queue->Owns(platform_queue, raster1_queue));
+  task_queue->Merge(platform_queue, raster2_queue);
+  ASSERT_TRUE(task_queue->Owns(platform_queue, raster2_queue));
+  const auto now = fml::TimePoint::Now();
+  int expected_value = 0;
+  // Right order:
+  // "test_val = 0" in raster1_queue
+  // "test_val = 1" in platform_queue
+  // "test_val = 2" in raster2_queue
+  for (;;) {
+    fml::closure invocation = task_queue->GetNextTaskToRun(platform_queue, now);
+    if (!invocation) {
+      break;
+    }
+    invocation();
+    ASSERT_TRUE(test_val == expected_value);
```

Comment:
1. could you also capture the queue_id that the invocation has run on, and assert that it is indeed the right queue?
2. mind adding a similar test that also tests unmerge in the following ways:
  a. Unmerge will not unmerge all the subsumed queues.
  b. Unmerge will respect the ordering of the tasks in all of the queues.
  c. Can not merge the same queue to 2 different owners.

Reply:
1. In this test, the invocation() is called in the test body and it does not really runs task queue on the children thread. There's a another test called `TEST(RasterThreadMerger, IsNotOnRasterizingThread) ` in `raster_thread_merger_unittests.cc` captured the queue_id and checked the right queue or not.
2. Some more test added for a, b, c three ways.
a. `OneCanOwnMultipleQueuesAndUnmergeIndependently` in `message_loop_task_queues_merge_unmerge_unittests.cc`
b.  `UnmergeRespectTheOriginalTaskOrderingInQueues` in `message_loop_task_queues_unittests.cc`
c. `CannotMergeSameQueueToTwoDifferentOwners` in `message_loop_task_queues_merge_unmerge_unittests.cc`

These advices really helpful for me. 👍

And the three files(message_loop_task_queues_merge_unmerge_unittests.cc, message_loop_task_queues_unittests.cc and raster_thread_merger_unittests.cc) is a little confusing and let us hard to distinguish them.